metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/usage.py

```diff
@@ -158,6 +170,147 @@ def bigquery_audit_metadata_query_template(
+
+    def __init__(self, config: BigQueryV2Config):
+        self.conn = ConnectionWrapper()
+        self.read_events = FileBackedDict[ReadEvent](
+            shared_connection=self.conn,
+            tablename="read_events",
+            extra_columns={
+                "resource": lambda e: str(e.resource),
+                "name": lambda e: e.jobName,
+                "timestamp": lambda e: get_time_bucket(
+                    e.timestamp, config.bucket_duration
+                ),
+                "user": lambda e: e.actor_email,
+            },
+            cache_max_size=config.file_backed_cache_size,
+        )
+        # Keyed by job_name
+        self.query_events = FileBackedDict[QueryEvent](
+            shared_connection=self.conn,
+            tablename="query_events",
+            extra_columns={
+                "query": lambda e: e.query,
+                "is_read": lambda e: int(e.statementType in READ_STATEMENT_TYPES),
+            },
+            cache_max_size=config.file_backed_cache_size,
+        )
+        # Created just to store column accesses in sqlite for JOIN
+        self.column_accesses = FileBackedDict[Tuple[str, str]](
+            shared_connection=self.conn,
+            tablename="column_accesses",
+            extra_columns={"read_event": lambda p: p[0], "field": lambda p: p[1]},
+            cache_max_size=config.file_backed_cache_size,
+        )
+
+    def close(self) -> None:
+        self.read_events.close()
+        self.query_events.close()
+        self.column_accesses.close()
+        self.conn.close()
+
+    def standalone_events(self) -> Iterable[AuditEvent]:
+        for read_event in self.read_events.values():
+            query_event = (
+                self.query_events.get(read_event.jobName)
+                if read_event.jobName
+                else None
+            )
+            yield AuditEvent(read_event=read_event, query_event=query_event)
+        for _, query_event in self.query_events.items_snapshot("NOT is_read"):
+            yield AuditEvent(query_event=query_event)
+
+    @staticmethod
+    def usage_statistics_query(top_n: int) -> str:
+        return f"""
+        SELECT a.timestamp, a.resource, a.query_count, b.query_freq, c.user_freq, d.column_freq FROM (
+            SELECT
+                r.timestamp,
+                r.resource,
+                COUNT(q.query) query_count
+            FROM
+                read_events r
+                LEFT JOIN query_events q ON r.name = q.key
+            GROUP BY r.timestamp, r.resource
+        ) a
+        LEFT JOIN (
+            SELECT timestamp, resource, json_group_array(json_array(query, query_freq)) as query_freq FROM (
+                SELECT
+                    r.timestamp,
+                    r.resource,
+                    q.query,
+                    COUNT(r.key) as query_freq,
+                    ROW_NUMBER() over (PARTITION BY r.timestamp, r.resource, q.query ORDER BY COUNT(r.key) DESC, q.query) as rank
```

Comment:
Interesting ! so here we are using GROUP BY and WINDOW FN in same query, also the grouping(partitioning) criterion is same, and seemingly it works. If we have tested this against real data, then this is fine.

Reply:
This has been tested against unit tests and performance tests, and I'll run it against connector tests soon. If all 3 pass, I feel good enough about this